### ISSUE TITLE
Type nullability is shaved off when using class templates

### ISSUE THREAD

**Issue by dko-slapdash** [Microsoft/TypeScript#31245]
**TypeScript Version:**
3.4.0-dev.201xxxxx

**Search Terms:**
typescript class template null disappears

**Code**

```ts
function tst(s: string) { }

class Cls<TField extends string> {
  constructor(private field: TField) { }

  fff(rec: Record<TField, string | null>) {
    const val = rec[this.field]; // Record<TField, string>[TField] - where is null?
    /*if (!val) {
      return;
    }*/
    tst(val); // must NOT work, but works
  }
}

const c = new Cls("aa");
c.fff({ aa: null });
```

**Expected behavior:**
It should complain that tst() can be called with a nullable value

**Actual behavior:**
It compiles. In TS 3.4.3 nullability was NOT removed, but if we uncomment the `if` statement in the example, an opposite behavior appears: `tst()` complains that val can be null, although we check & early-return it above.

**Playground Link:**
[playground link](https://www.typescriptlang.org/play/#src=const%20SPECIAL%3A%20unique%20symbol%20%3D%20Symbol(%22sym%22)%3B%0D%0A%0D%0Aabstract%20class%20Base%20%7B%0D%0A%20%20abstract%20fff()%3A%20typeof%20SPECIAL%20%7C%20number%3B%0D%0A%7D%0D%0A%0D%0Aclass%20Derive1%20extends%20Base%20%7B%0D%0A%20%20fff()%20%7B%0D%0A%20%20%20%20return%2010%3B%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0Aclass%20Derive2%20extends%20Base%20%7B%0D%0A%20%20fff()%20%7B%0D%0A%20%20%20%20%2F%2F%20Error%3A%20Type%20'symbol'%20is%20not%20assignable%20to%20type%20'number%20%7C%20unique%20symbol'%0D%0A%20%20%20%20return%20SPECIAL%3B%0D%0A%20%20%7D%0D%0A%7D%0D%0A)

**Related Issues:**

**Comment by weswigham** [comment]
You need to enable `strictNullChecks`, or the parent flag `strict`. It works just dandy on the [playground](https://www.typescriptlang.org/play/#src=function%20tst(s%3A%20string)%20%7B%20%7D%0D%0A%0D%0Aclass%20Cls%3CTField%20extends%20string%3E%20%7B%0D%0A%20%20constructor(private%20field%3A%20TField)%20%7B%20%7D%0D%0A%0D%0A%20%20fff(rec%3A%20Record%3CTField%2C%20string%20%7C%20null%3E)%20%7B%0D%0A%20%20%20%20const%20val%20%3D%20rec%5Bthis.field%5D%3B%20%2F%2F%20Record%3CTField%2C%20string%3E%5BTField%5D%20-%20where%20is%20null%3F%0D%0A%20%20%20%20%2F*if%20(!val)%20%7B%0D%0A%20%20%20%20%20%20return%3B%0D%0A%20%20%20%20%7D*%2F%0D%0A%20%20%20%20tst(val)%3B%20%2F%2F%20must%20NOT%20work%2C%20but%20works%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0Aconst%20c%20%3D%20new%20Cls(%22aa%22)%3B%0D%0Ac.fff(%7B%20aa%3A%20null%20%7D)%3B) once you do as much.

**Comment by jcalz** [comment]
Please recheck your code with the [`--strictNullChecks` compiler option](https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#--strictnullchecks) enabled, and please check your playground link to make sure the code is relevant to this issue.  Thanks.

**Comment by dko-slapdash** [comment]
Oh, the issue is in the commented code. I'm closing the current issue as unrelated to the problem and opening a clean new one with details: https://github.com/Microsoft/TypeScript/issues/31274
